More user-friendly message when calling minimum_of()/maximum_of() with no args

If you try to write `minimum_of()` or `maximum_of()` in a dataset definition, with no arguments, you get a horrible error message raised from the type checking when the functions attempt to cast their arguments.  Failing to supply arguments is an error, but we could report it in a more comprehensible way.

```
    minimum_of(),
    ^^^^^^^^^^^^
  File "/home/becky/datalab/ehrql/ehrql/query_language.py", line 1340, in minimum_of
    return _apply(qm.Function.MinimumOf, args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/becky/datalab/ehrql/ehrql/query_language.py", line 923, in _apply
    return _wrap(qm_node)
           ^^^^^^^^^^^^^^
  File "/home/becky/datalab/ehrql/ehrql/query_language.py", line 894, in _wrap
    type_ = get_series_type(qm_node)
            ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/becky/datalab/ehrql/ehrql/query_model/nodes.py", line 453, in get_series_type
    typespec = get_typespec(series)
               ^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/functools.py", line 909, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/becky/datalab/ehrql/ehrql/query_model/nodes.py", line 777, in get_typespec_for_series
    typevar_value = resolve_typevar_from_inputs(series, typevar)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/becky/datalab/ehrql/ehrql/query_model/nodes.py", line 846, in resolve_typevar_from_inputs
    assert False, f"Could not match TypeVar {typevar} in {series}"
AssertionError: Could not match TypeVar ~Comparable in Function.MinimumOf(sources=())
```

--- a/docs/includes/generated_docs/language__functions.md
+++ b/docs/includes/generated_docs/language__functions.md
@@ -44,7 +44,7 @@ category = when(size < 15).then("small").otherwise("large")
 
 
 <h4 class="attr-heading" id="maximum_of" data-toc-label="maximum_of" markdown>
-  <tt><strong>maximum_of</strong>(<em>*args</em>)</tt>
+  <tt><strong>maximum_of</strong>(<em>value</em>, <em>other_value</em>, <em>*other_values</em>)</tt>
 </h4>
 <div markdown="block" class="indent">
 Return the maximum value of a collection of Series or Values, disregarding NULLs
@@ -58,7 +58,7 @@ latest_event_date = maximum_of(event_series_1.date, event_series_2.date, "2001-0
 
 
 <h4 class="attr-heading" id="minimum_of" data-toc-label="minimum_of" markdown>
-  <tt><strong>minimum_of</strong>(<em>*args</em>)</tt>
+  <tt><strong>minimum_of</strong>(<em>value</em>, <em>other_value</em>, <em>*other_values</em>)</tt>
 </h4>
 <div markdown="block" class="indent">
 Return the minimum value of a collection of Series or Values, disregarding NULLs

--- a/docs/includes/generated_docs/specs.md
+++ b/docs/includes/generated_docs/specs.md
@@ -1752,52 +1752,6 @@ returns the following patient series:
 
 
 
-#### 6.5.20 Maximum of single value
-
-This example makes use of a patient-level table named `p` containing the following data:
-
-| patient|i1|i2|d1|d2|s1|s2|f1|f2 |
-| - | - | - | - | - | - | - | - | - |
-| 1|101|112|2001-01-01|2012-12-12|a|d|1.01|1.12 |
-| 2||211||2021-01-01||f||2.11 |
-| 3|||||||| |
-
-```python
-maximum_of(0)
-```
-returns the following patient series:
-
-| patient | value |
-| - | - |
-| 1|0 |
-| 2|0 |
-| 3|0 |
-
-
-
-#### 6.5.21 Minimum of single value
-
-This example makes use of a patient-level table named `p` containing the following data:
-
-| patient|i1|i2|d1|d2|s1|s2|f1|f2 |
-| - | - | - | - | - | - | - | - | - |
-| 1|101|112|2001-01-01|2012-12-12|a|d|1.01|1.12 |
-| 2||211||2021-01-01||f||2.11 |
-| 3|||||||| |
-
-```python
-minimum_of(0)
-```
-returns the following patient series:
-
-| patient | value |
-| - | - |
-| 1|0 |
-| 2|0 |
-| 3|0 |
-
-
-
 ### 6.6 Minimum and maximum aggregations across Event series
 
 

--- a/ehrql/query_language.py
+++ b/ehrql/query_language.py
@@ -1350,7 +1350,7 @@ def case(*when_thens, default=None):
 
 # HORIZONTAL AGGREGATION FUNCTIONS
 #
-def maximum_of(*args):
+def maximum_of(value, other_value, *other_values):
     """
     Return the maximum value of a collection of Series or Values, disregarding NULLs
 
@@ -1359,11 +1359,11 @@ def maximum_of(*args):
     latest_event_date = maximum_of(event_series_1.date, event_series_2.date, "2001-01-01")
     ```
     """
-    args = cast_all_arguments(args)
+    args = cast_all_arguments((value, other_value, *other_values))
     return _apply(qm.Function.MaximumOf, args)
 
 
-def minimum_of(*args):
+def minimum_of(value, other_value, *other_values):
     """
     Return the minimum value of a collection of Series or Values, disregarding NULLs
 
@@ -1372,5 +1372,5 @@ def minimum_of(*args):
     ealiest_event_date = minimum_of(event_series_1.date, event_series_2.date, "2001-01-01")
     ```
     """
-    args = cast_all_arguments(args)
+    args = cast_all_arguments((value, other_value, *other_values))
     return _apply(qm.Function.MinimumOf, args)

--- a/tests/integration/test_query_engines.py
+++ b/tests/integration/test_query_engines.py
@@ -6,7 +6,7 @@ import sqlalchemy
 
 from ehrql import Dataset
 from ehrql.query_language import PatientFrame, Series, table_from_file, table_from_rows
-from ehrql.query_model.nodes import Value
+from ehrql.query_model.nodes import Function, Value
 from ehrql.tables.beta.core import clinical_events, patients
 
 
@@ -153,3 +153,40 @@ def test_cleans_up_temporary_tables(engine):
 def _get_tables(engine):
     inspector = sqlalchemy.inspect(engine.sqlalchemy_engine())
     return sorted(inspector.get_table_names())
+
+
+# Supplying only a single series to the min/max functions is valid in the query model so
+# Hypothesis will generate examples like this which we want to handle correctly. But we
+# deliberately make these unconstructable in ehrQL so we can't write standard spec tests
+# to cover them.
+@pytest.mark.parametrize(
+    "operation",
+    [
+        Function.MinimumOf,
+        Function.MaximumOf,
+    ],
+)
+def test_minimum_maximum_of_single_series(engine, operation):
+    engine.populate(
+        {
+            patients: [
+                dict(patient_id=1, date_of_birth=date(1980, 1, 1)),
+                dict(patient_id=2, date_of_birth=date(1990, 2, 2)),
+            ]
+        }
+    )
+
+    variables = dict(
+        population=as_query_model(patients.exists_for_patient()),
+        v=operation(
+            (as_query_model(patients.date_of_birth),),
+        ),
+    )
+    assert engine.extract_qm(variables) == [
+        {"patient_id": 1, "v": date(1980, 1, 1)},
+        {"patient_id": 2, "v": date(1990, 2, 2)},
+    ]
+
+
+def as_query_model(query_lang_expr):
+    return query_lang_expr._qm_node

--- a/tests/spec/series_ops/test_maximum_of_and_minimum_of_patient_series.py
+++ b/tests/spec/series_ops/test_maximum_of_and_minimum_of_patient_series.py
@@ -244,27 +244,3 @@ def test_maximum_of_two_integers_all_a_values(spec_test):
             3: 3,
         },
     )
-
-
-def test_maximum_of_single_value(spec_test):
-    spec_test(
-        table_data,
-        maximum_of(0),
-        {
-            1: 0,
-            2: 0,
-            3: 0,
-        },
-    )
-
-
-def test_minimum_of_single_value(spec_test):
-    spec_test(
-        table_data,
-        minimum_of(0),
-        {
-            1: 0,
-            2: 0,
-            3: 0,
-        },
-    )

